### DISCUSSION TITLE
Is it possible to subscribe query changes in a separate component?

### DISCUSSION THREAD

**Discussion by baymac** [TanStack/query#846] [General] [answered]
Using `queryCache.getQueryData(...)` does not get updated when the query if fetched, it is only updated once the component renders for some other state change. It would be nice to be able to subscribe changes like a state.

**Comment by Jaypov** [comment]
I am wondering the same thing, this would be very powerful if its possible. However from reading the docs it seems that it is not.

I assumed subscribing to an observer with `queryCache.subscribe` would work but this function states

> Out of scope mutations to the queryCache are not encouraged and will not fire subscription callbacks

if I am correct what is the reasoning for this?

**Comment by TkDodo** [comment, marked as answer]
I think you would just do `useQuery` _again_ in that other component? Or am I missing something here?

**Reply by Basovs** [reply]
useQuery makes a call to the API. If you use 'useQuery' in several components - it will lead to that you fetch same data several times... Would be nice to have a global reactive 'data' state for a specific query key.

**Reply by TkDodo** [reply]
It will not lead to several fetches if you set a `staleTime`.

**Reply by mcdougald** [reply]
> I think you would just do `useQuery` _again_ in that other component? Or am I missing something here?

Hi TkDodo, thanks for the library

My use case is similar to https://github.com/TanStack/query/discussions/3756. I have a list of 100 items, each needs to useQuery 1 - 7 times. All the Observers (and I guess timeouts?) are causing the page to lag. 

So I've lifted the data/errors for each useQuery by creating a context for it.

But React's context isn't the greatest solution either right? So I've been looking at other options too. getQueryData, getQueryState, etc... but those aren't updating with results from the latest calls.

I believe I saw in a diff discussion Tanner saying react-query's context can be used for adv state management (like this)? But I can't find a clear setup/example. Do I need an observer in each item no matter what for isLoading, isFetched, etc?

**Reply by TkDodo** [reply]
We had the same question / discussion in discord lately (maybe it was even you 😅 ?) So I'll just re-post my answer here for posterity:

the problem with getQueryData is that it does not create a subscription. So if the query changes due to a background refetch, those components won't be informed about that change directly. Sure, if the parent re-renders, they will pick it up, but if the parent re-renders anyways, you can just as well stick the user data in context, which will at least inform all consumers.

of course, you loose the ability of fine-grained subscriptions via select (when using context), where you could subscribe to one property of a query result (e.g. email of a user) only, and only re-render if that email changes. It's a tradeoff.

**Reply by shakedlokits** [reply]
I only have a partial key in different component, i.e I fetched the bill with id=3 with `['bill', 3]` but I want to know in a different component(not fetch) if all of the bills finished fetching like `['bill']`.

**Reply by tholder** [reply]
I'm adding this here because we had a real problem with useQuery seemingly not working as indicated here. Changes were not being updated in sibling components as no reaction was occurring. Previously we had solved this by loading in to state higher up and prop-drilling it down through components. Obviously this is a pain. Turns out that setting notifyOnChangeProps property to 'all' fixes the issue for us and the following post was helpful https://tkdodo.eu/blog/react-query-render-optimizations - I don't fully understand the implication of this but for us it solves the issue.

**Reply by TkDodo** [reply]
I can take a look into the issue if you can show a reproduction where it fails please.

**Comment by baymac** [comment]
You can use `useQueryCache` hook. E.g. Inside your FC:

```
const queryCache = useQueryCache()
const item = queryCache.getQueryData([key, params])
```

Now your component will subscribe to the changes in queryCache and item will be updated.

**Reply by TkDodo** [reply]
Are you certain of that? To my knowledge, the queryCache is just made available via react context - `useQueryCache` is just a wrapper around that. The cache itself is stable, so it never changes, and `getQueryData` doesn't subscribe you....

**Reply by nilavrusty** [reply]
@TkDodo Could you let me know how may I subscribe to only a particular queryKey instead of subscribing to the complete queryCache

**Reply by TkDodo** [reply]
That is what `useQuery` is doing ...

**Reply by nilavrusty** [reply]
@TkDodo  useQuery would need queryFn, i wanted something like subscribe(myqueryKey),

**Reply by TkDodo** [reply]
I'm not sure that's possible. Maybe by creating your own `Observer` ... I'm not sure what problem you are trying to solve with this though. You can achieve a similar behaviour by just calling `useQuery`, best abstracted in a custom hook, and if you don't want background refetches, you can customize `staleTime`.

**Reply by araskavakli** [reply]
> you actually can set the `enabled` option to false and then the `queryFn` would no longer be required

Nice. Thanks!

**Comment by sergiodxa** [comment]
You can use [`queryCache.subscribe`](https://react-query.tanstack.com/docs/api#querycachesubscribe), but the easiest and probably safer way is to call useQuery again, it will not trigger a new render and you will know when something change.

**Comment by retyui** [comment]
```tsx
queryClient.getQueryCache().subscribe((queryCacheNotifyEvent) => {
  if (!queryCacheNotifyEvent) {
    return;
  }

  console.log(queryCacheNotifyEvent);
});
```

**Comment by saeedghasemi72** [comment]
I think QueryObserver can help you 

const observer = new QueryObserver(queryClient, { queryKey: ['posts'] })

const unsubscribe = observer.subscribe(result => {
  console.log(result)
  unsubscribe()
})

**Reply by pencilcheck** [reply]
observer seems to mess up other useQuery instance on other components... not working.

**Reply by alexmccabe** [reply]
I too tried this and it basically creates a new `useQuery` with the query key, and overrides any `useQuery` options you've added, so `staleTime` gets reset to default

**Comment by itsramiel** [comment]
Pass the key but not the function like this: 

```
const { data } = useQuery<T>([key]);
```

**Reply by Jaggler3** [reply]
It gives `WARN  Missing queryFn`

**Reply by ovidijusr** [reply]
Same to me, any ideas why? Also that query that was loading just fails to load

**Comment by danvc** [comment]
I was facing the same problem, here's what I did and it's working fine, as expected:

To listen to the query:

```
  const { data: dateRange } = useQuery({
    enabled: true,
    queryKey: ["queryName"],
    queryFn: () => null,
  });
  console.log("dateRange", dateRange);
```

**Reply by nadavge** [reply]
As far as I understand, this will set the value of ["queryName"] in the cache to `null` which I assume you don't want to do...

**Comment by venn24** [comment]
i use useMutationState

**Comment by dberardo-com** [comment]
how about 

queryClient.getQueryCache().subscribe((queryCacheNotifyEvent) => {
                          let hash = queryCacheNotifyEvent.query?.queryHash as any;

... check the hash and do stuff
                        });

**Reply by alexmccabe** [reply]
This event fires every single time the cache is changed, and invalidating a query caused the callback to trigger six times, not just when data changes. You'd have to do deep comparison to make sure the data had changed, and that could be a performance issue

**Comment by alexmccabe** [comment]
For me this would be useful as I have some local state that needs to be updated with the partial result of a query, so I need to know _when_ that has happened to update the partial state. I'm using vue-query, and I'm trying to avoid placing a watcher on the result and instead subscribe to it. 

I had good results with the QueryObserver, however I don't like how I have to pass all the exact same options to it otherwise the initial query config gets overridden, and this is going to be difficult for the rest of the team to remember to do, which will eventually introduce bugs.

**Reply by ViewableGravy** [reply]
I have only used `useQuery` in react, but the general concept should be the same across frameworks.

The recommended approach to creating queries is as follows:
```ts
// create one of these on a per endpoint/async function basis typically
const createQueryAOptions = (opts: MyOpts) => {
  return queryOptions({
    queryKey: ["queryA", opts],
    queryFn: () => myAPI(opts),
    ...custom opts,
  }) 
}

// In your component
const { data } = useQuery(createQueryAOptions(opts));
```

In your case, since you want to derive some state, you would want a select function:
```ts
const { data } = useQuery({
  ...createQueryAOptions(opts),
  select: (data) => ...
});
```

This will not refetch using the queryFn after the first run, since the query is cached, and if you do an optimistic update on the query, your useQuery will re-run the select and re-render your component accordingly.

**Comment by alexmccabe** [comment]
Here's where I've ended up. Obviously this relies on Vue internals, but I've also added a `select` option, to do some data narrowing like `useQuery` allows. Could change the API to be a singular object instead, for consistency if desired

```ts
export function useQuerySubscriber<TQueryFnData = unknown, TError = Error, TData = TQueryFnData>(
  queryKey: QueryKey,
  callback: (
    state: QueryState<TData, TError> | undefined,
    previousState: QueryState<TData, TError> | undefined
  ) => void,
  options: { select?: (data: TQueryFnData | undefined) => TData | undefined } = {}
) {
  const state = ref<QueryState<TQueryFnData, TError> | undefined>();
  const client = useQueryClient();

  const unsubscribe = client.getQueryCache().subscribe(() => {
    const newValue = client.getQueryState<TQueryFnData, TError>(queryKey);

    if (options.select && newValue?.data) {
      return callback(
        { ...newValue, data: options.select(newValue.data) },
        state.value ? { ...state.value, data: options.select(state.value?.data) } : undefined
      );
    }

    callback(newValue as unknown as QueryState<TData, TError>, state.value as unknown as QueryState<TData, TError>);
  });

  onScopeDispose(() => unsubscribe());
}
```
**Usage:**
```ts
useQuerySubscriber<{ data: DuplicateContactsParams }, Error, DuplicateContactsParams>(
  getDuplicateContactsParamsQueryKey,
  (newState) => {
    if (!newState?.data) return;

    insightParams.value = {
      ...insightParams.value,
      activityStart: newState.data.activityStart,
      precision: newState.data.precision,
    };
  },
  { select: (state) => state?.data }
);
```

**Comment by abielzulio** [comment]
you can achieve this functionality using `React.useSyncExternalStore` easily

```typescript

import { useQueryClient } from "@tanstack/react-query"
import React from "react"

export function useQueryCacheState<T>(
  key: readonly unknown[],
  initialData: T
): [T, (updater: T | ((prev: T) => T)) => void] {
  const queryClient = useQueryClient()

  const cache = queryClient.getQueryCache()

  const getSnapshot = () => {
    return queryClient.getQueryData<T>(key) ?? initialData
  }

  const subscribe = (callback: () => void) => {
    return cache.subscribe((event) => {
      if (
        event.type === "updated" &&
        JSON.stringify(event.query.queryKey) === JSON.stringify(key)
      ) {
        callback()
      }
    })
  }

  const data = React.useSyncExternalStore(subscribe, getSnapshot, getSnapshot)

  const setData = (updater: T | ((prev: T) => T)) => {
    queryClient.setQueryData<T>(
      key,
      typeof updater === "function"
        ? (prev) => (updater as (prev: T) => T)(prev ?? initialData)
        : () => updater
    )
  }

  return [data, setData]
}
```
